core/src/main/scala/org/bitcoins/core/wallet/signer/Signer.scala

```diff
@@ -385,6 +368,70 @@ sealed abstract class MultiSigSigner
 
 object MultiSigSigner extends MultiSigSigner
 
+/** Used to sign a [[org.bitcoins.core.protocol.script.P2SHScriptPubKey]] */
+sealed abstract class P2SHSigner extends BitcoinSigner[P2SHSpendingInfo] {
+  override def sign(
+      spendingInfo: UTXOSpendingInfo,
+      unsignedTx: Transaction,
+      isDummySignature: Boolean,
+      spendingInfoToSatisfy: P2SHSpendingInfo)(
+      implicit ec: ExecutionContext): Future[TxSigComponent] = {
+    if (spendingInfoToSatisfy != spendingInfo) {
+      Future.fromTry(TxBuilderError.WrongSigner)
+    } else {
+
+      val (_, output, inputIndex, _) =
+        relevantInfo(spendingInfo, unsignedTx)
+
+      if (spendingInfoToSatisfy
+            .isInstanceOf[P2SHNestedSegwitV0UTXOSpendingInfo] && spendingInfoToSatisfy.redeemScript
+            .isInstanceOf[P2WPKHWitnessSPKV0]) {
+        val p2wpkh =
+          spendingInfoToSatisfy.redeemScript.asInstanceOf[P2WPKHWitnessSPKV0]
+        val pubKey = spendingInfo.signers.head.publicKey
+        if (P2WPKHWitnessSPKV0(pubKey) != p2wpkh) {
+          Future.fromTry(TxBuilderError.WrongPublicKey)
+        } else {
+          val p2shScriptSig = P2SHScriptSignature(p2wpkh)
+          val oldInput = unsignedTx.inputs(inputIndex.toInt)
+          val updatedInput = TransactionInput(oldInput.previousOutput,
+                                              p2shScriptSig,
+                                              oldInput.sequence)
+
+          val uwtx = {
+            val u = unsignedTx.updateInput(inputIndex.toInt, updatedInput)
+            WitnessTransaction.toWitnessTx(u)
+          }
+
+          Future.successful(
+            WitnessTxSigComponentP2SH(transaction = uwtx,
+                                      inputIndex = inputIndex,
+                                      output = output,
+                                      flags = Policy.standardFlags))
+        }
+      } else {
+        val signedSigComponentF = BitcoinSigner.sign(
+          spendingInfoToSatisfy.nestedSpendingInfo,
+          unsignedTx,
+          isDummySignature,
+          spendingInfoToSatisfy.nestedSpendingInfo)
+
+        val scriptSigF = signedSigComponentF.map { signedSigComponent =>
+          P2SHScriptSignature(signedSigComponent.scriptSignature,
+                              spendingInfoToSatisfy.redeemScript)
+        }
+
+        updateScriptSigInSigComponent(unsignedTx,
+                                      inputIndex.toInt,
+                                      output,
+                                      scriptSigF)
+      }
```

Comment:
I'm still not sure this covers this case either, take a look at `TxBuilder.signAndAddInputForP2SH` and ditto the comment above about replacing stuff in `TxBuilder` for test coverage